### PR TITLE
DENG1546 -Add a derived dataset for serp events

### PR DESCRIPTION
**Pull request by alekhyamoz**
Reverts mozilla/bigquery-etl#4406

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-1744)

### CODE REVIEW

**Comment by alekhyamoz** [comment]
@dzeber -  included the missing updates here. You can close PR-4405 as it's no longer required.

**Comment by alekhyamoz** [comment]
> @alekhyamoz do we need to do anything special with the scheduling here, since the table should fill with an extra day's lag? Eg. when the job runs on 2023-10-11, the query will read data from both 2023-10-09 and 2023-10-10, and then write rows for 2023-10-09 only.

I think this is handled in the query itself 
 ```
AND DATE(submission_timestamp) >= DATE_SUB(@submission_date, INTERVAL 1 DAY)
    AND DATE_DIFF(DATE(submission_timestamp), @submission_date, DAY) IN (0, 1)
```

File: sql_generators/serp_events/templates/desktop_query.sql [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    event
+  FROM
+    `moz-fx-data-shared-prod.firefox_desktop_stable.events_v1`,
+    UNNEST(events) AS event
+  WHERE
+    event.category = 'serp'
+    -- allow events related to an impression_id to span 2 submission dates
+    -- we restrict to event sequences started on a single date below
+    AND DATE(submission_timestamp) >= DATE_SUB(@submission_date, INTERVAL 1 DAY)
+    AND DATE_DIFF(DATE(submission_timestamp), @submission_date, DAY) IN (0, 1)
```
Comment (dzeber) [inline sql_generators/serp_events/templates/desktop_query.sql:14]: IIUC `@submission_date` is the most recent date with complete data? In that case, we want this query to cover `@submission_date` and the day before `@submission_date`.

So this should be:
```
AND DATE(submission_timestamp) >= DATE_SUB(@submission_date, INTERVAL 1 DAY)
AND DATE(submission_timestamp) <= @submission_date
```

File: sql_generators/serp_events/templates/desktop_query.sql [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    CAST(mozfun.map.get_key(event.extra, 'is_shopping_page') AS bool) AS is_shopping_page,
+    mozfun.map.get_key(event.extra, 'provider') AS search_engine,
+    mozfun.map.get_key(event.extra, 'source') AS sap_source,
+    CAST(mozfun.map.get_key(event.extra, 'tagged') AS bool) AS is_tagged
+  FROM
+    serp_events
+  WHERE
+    event.name = 'impression'
+    -- restrict to sessions that started on the target submission date
+    AND DATE(submission_timestamp) = @submission_date
```
Comment (dzeber) [inline sql_generators/serp_events/templates/desktop_query.sql:72]: And then this should be  `AND DATE(submission_timestamp) = DATE_SUB(@submission_date, INTERVAL 1 DAY)`?

This CTE determines the date for the final rows that will be written by this run. The query should write rows for the day before the latest date with complete data.

**Review by dzeber — COMMENTED** [review]
@alekhyamoz thanks! I want to make sure I understand how `@submission_date` is used - if it's the latest date with complete data, then I think the filters need to be updated as I've commented.

**Review by dzeber — APPROVED** [review]
Looks great, thanks!

**Pull request merged**